stream.go

```diff
@@ -525,27 +525,33 @@ func (cs *clientStream) commitAttempt() {
 // shouldRetry returns nil if the RPC should be retried; otherwise it returns
 // the error that should be returned by the operation.
 func (cs *clientStream) shouldRetry(err error) error {
-	unprocessed := false
 	if cs.attempt.s == nil {
+		// Error from NewClientStream.
 		pioErr, ok := err.(transport.PerformedIOError)
-		if ok {
-			// Unwrap error.
-			err = toRPCErr(pioErr.Err)
-		} else {
-			unprocessed = true
-		}
-		if !ok && !cs.callInfo.failFast {
+		if !ok {
 			// In the event of a non-IO operation error from NewStream, we
 			// never attempted to write anything to the wire, so we can retry
-			// indefinitely for non-fail-fast RPCs.
+			// indefinitely.  Except for INTERNAL errors, which indicate the
+			// RPC should not be retried due to max header list size violation.
+			if status.Convert(err).Code() == codes.Internal {
```

Comment:
I'm slightly concerned about this check for `codes.Internal`. Would there be an `Internal` error that's retriable? Should we make another export type to wrap this? `DoNotRetryThisItWillFailAgainError`

Reply:
Fixed.  The flow around error handling is pretty complex because of the need to wrap & unwrap errors.  I thought about implementing `Unwrap() error` and also implementing `GRPCStatus()` on `NewStreamError` but I think this is okay.  LMK what you think.